### PR TITLE
[clang] static operators should evaluate object argument (reland)

### LINKED ISSUE #67976: Too aggresive C++23 static call operator optimization
Hello,

It looks like clang is too aggressive for C++23 static call optimization. Here is the simplified example of real code:

```cpp
#include <functional>
#include <iostream>

template <typename Tag>
struct Wrapper {
    template <typename F, typename... Args>
    static void operator()(F&& callable, Args&&... args) {
        Invoke(std::forward<F>(callable), std::forward<Args>(args)...);
    }

    template <typename F, typename... Args>
    static void Invoke(F&& callable, Args&&... args) {
        std::invoke(std::forward<F>(callable), std::forward<Args>(args)...);
    }
};

struct Manager {
    template <typename SingletonClass>
    static SingletonClass& Get() {
        std::cout << "get call\n";
        static SingletonClass instance;
        return instance;
    }
};

struct Tag {};

int main()
{
    using TaggedWrapper = Wrapper<Tag>;
    Manager::Get<TaggedWrapper>()([] { std::cout << "Function\n"; });
}
```

We assume that Manager::Get() will be called and `get call` will be printed, but it doesn't happen. If change call in main to `    Manager::Get<TaggedWrapper>().Invoke([] { std::cout << "Function\n"; });` it works as expected. `get cal` also will be printed if remove `static` from `operator()`. It reproducible even with `-O0`. clang-17.0.1

### PR DESCRIPTION
**Pull request by SuperSodaSea**
### Description

clang don't evaluate the object argument of `static operator()` and `static operator[]` currently, for example:

```cpp
#include <iostream>

struct Foo {
    static int operator()(int x, int y) {
        std::cout << "Foo::operator()" << std::endl;
        return x + y;
    }
    static int operator[](int x, int y) {
        std::cout << "Foo::operator[]" << std::endl;
        return x + y;
    }
};
Foo getFoo() {
    std::cout << "getFoo()" << std::endl;
    return {};
}
int main() {
    std::cout << getFoo()(1, 2) << std::endl;
    std::cout << getFoo()[1, 2] << std::endl;
}
```

`getFoo()` is expected to be called, but clang don't call it currently (17.0.6). This PR fixes this issue.

Fixes #67976, reland #68485.

### Walkthrough

- **clang/lib/Sema/SemaOverload.cpp**
  - **`Sema::CreateOverloadedArraySubscriptExpr` & `Sema::BuildCallToObjectOfClassType`**
  Previously clang generate `CallExpr` for static operators, ignoring the object argument. In this PR `CXXOperatorCallExpr` is generated for static operators instead, with the object argument as the first argument.
  - **`TryObjectArgumentInitialization`**
  `const` / `volatile` objects are allowed for static methods, so that we can call static operators on them.
- **clang/lib/CodeGen/CGExpr.cpp**
  - **`CodeGenFunction::EmitCall`**
  CodeGen changes for `CXXOperatorCallExpr` with static operators: emit and ignore the object argument first, then emit the operator call.
- **clang/lib/AST/ExprConstant.cpp**
  - **`‎ExprEvaluatorBase::handleCallExpr‎`**
  Evaluation of static operators in constexpr also need some small changes to work, so that the arguments won't be out of position.
- **clang/lib/Sema/SemaChecking.cpp**
  - **`Sema::CheckFunctionCall`**
  Code for argument checking also need to be modify, or it will fail the test `clang/test/SemaCXX/overloaded-operator-decl.cpp`.
- **clang-tools-extra/clangd/InlayHints.cpp**
  - **`InlayHintVisitor::VisitCallExpr`**
  Now that the `CXXOperatorCallExpr` for static operators also have object argument, we should also take care of this situation in clangd.

### Tests

- **Added:**
    - **clang/test/AST/ast-dump-static-operators.cpp**
      Verify the AST generated for static operators.
    - **clang/test/SemaCXX/cxx2b-static-operator.cpp**
      Static operators should be able to be called on const / volatile objects.
- **Modified:**
    - **clang/test/CodeGenCXX/cxx2b-static-call-operator.cpp**
    - **clang/test/CodeGenCXX/cxx2b-static-subscript-operator.cpp**
      Matching the new CodeGen.

### Documentation

- **clang/docs/ReleaseNotes.rst**
  Update release notes.

### CODE REVIEW

**Comment by SuperSodaSea** [comment]
CC @cor3ntin, @shafik, @AaronBallman and @zyn0217

File: clang-tools-extra/clangd/InlayHints.cpp [hunk 1]
```diff
@@ -651,16 +651,12 @@ class InlayHintVisitor : public RecursiveASTVisitor<InlayHintVisitor> {
...
-    // However, we don't have the implied object argument
-    // for static operator() per clang::Sema::BuildCallToObjectOfClassType.
     llvm::ArrayRef<const Expr *> Args = {E->getArgs(), E->getNumArgs()};
     // We don't have the implied object argument through a function pointer
     // either.
     if (const CXXMethodDecl *Method =
             dyn_cast_or_null<CXXMethodDecl>(Callee.Decl))
-      if (Method->isInstance() &&
-          (IsFunctor || Method->hasCXXExplicitFunctionObjectParameter()))
+      if (IsFunctor || Method->hasCXXExplicitFunctionObjectParameter())
```
Comment (cor3ntin) [inline clang-tools-extra/clangd/InlayHints.cpp:659]: I don;t understand where the change is. `isInstance` is implied by `hasCXXExplicitFunctionObjectParameter` so isn't that change doing... nothing?

**Review by zyn0217 — APPROVED** [review]
Thanks again for working on it. The clangd part LGTM, and since the other part was left as-is and had been accepted in our previous attempt #68485, I believe we can now reland the patch.

Please keep an eye on the CI in case we have missed anything else.

File: clang-tools-extra/clangd/InlayHints.cpp (hunk 1 above)
Comment (SuperSodaSea) [reply clang-tools-extra/clangd/InlayHints.cpp:659]: `IsFunctor` can be true when `isInstance()` is false, this is the situation (static functor) we are dealing with.

Comment (zyn0217) [reply clang-tools-extra/clangd/InlayHints.cpp:659]: See https://github.com/llvm/llvm-project/pull/68485/files#diff-19c518dbc68b30c66e1a2b6bd523c005fb2050dcf1a0e92305df7ab3e1b9e9f3L15744-L15751. The patch actually changes the behavior when we put an implied object argument. (I was replying at wrong place in that PR...)

Comment (zyn0217) [reply clang-tools-extra/clangd/InlayHints.cpp:659]: So we may have an implied object argument for *static* operator() calls without an explicit object parameter, while it was only applied previously to *non-static* (`isInstance()`) operator() calls.

**Pull request merged**